_posts/2020-02-13-spring-boot-application-events-explained.md

```diff
@@ -0,0 +1,263 @@
+
+## Creating an `ApplicationEvent`
+
+We can publish application events using the Spring Framework’s event publishing mechanism. Prior to Spring 4.2, events had to extend `ApplicationEvent`. In latest versions, we can just publish any object as an event.
+We are passing `source` object on which the event initially occurred or with which the event is associated (never null) to `super`. We do this for event sourcing. In event sourcing, we capture the state of entity or aggregate as a sequence of state changing events.
+
+Let's create a custom event called `UserCreatedEvent` by extending `ApplicationEvent`
+
+```java
+class UserCreatedEvent extends ApplicationEvent {
+	private String name;
+
+	UserCreatedEvent(Object source, String name) {
+		super(source);
+		this.name = name;
+	}
+	...
+}
+```
+
+Event can also be a normal object which does not extend `ApplicationEvent`,
+
+```java
+class UserRemovedEvent {
+	private String name;
+
+	UserRemovedEvent(String name) {
+		this.name = name;
+	}
+	...
+}
+```
+
+## Publishing an `ApplicationEvent`
+
+Spring gives the flexibility to publish an arbitrary event and doesn't force us to extend from `ApplicationEvent`. The `ApplicationEventPublisher` interface has been extended to allow you to publish any object.
+
+Now that we have a `UserCreatedEvent` by extending `ApplicationEvent` and `UserRemovedEvent` without extending. When the object is not an `ApplicationEvent`, Spring will wrap it in a `PayloadApplicationEvent` for us. In order to publish it to its respective listeners, we need an instance of`ApplicationEventPublisher`:
+
+```java
+@Component
+class Publisher {
+	@Autowired
+	private ApplicationEventPublisher publisher;
+
+	void publishEvent(final String name) {
+		// Publishing event created by extending ApplicationEvent
+		publisher.publishEvent(new UserCreatedEvent(this, name));
+		// Publishing a object as an event
+		publisher.publishEvent(new UserRemovedEvent(name));
+	}
+}
+```
+
+## Listening to an Application Event
+
+Now that we know how to create and publish a custom event, let's see how we can listen to the event. An event can have multiple listeners doing different work based on the application requirements.
+
+There are two ways to define a listener. We can either use the `@EventListener` annotation or implement the `ApplicationListener` interface. In either case, the listener class has to be managed by Spring.
+
+### Annotation-Driven
+
+Starting with Spring 4.1 it's now possible to simply annotate a method of a managed bean with `@EventListener` to automatically register an `ApplicationListener` matching the signature of the method. No additional configuration is necessary with annotation-driven configuration enabled. If your method should listen to several events or if you want to define it with no parameter at all, the event types can also be specified on the annotation itself. Example: `@EventListener({ContextStartedEvent.class, ContextRefreshedEvent.class})`.
```

Comment:
This paragraph is again explaining code before the reader had a chance to look at the code example. Please re-structure.